From d3ece29f674b85c055e15b955adc867e91e6b25e Mon Sep 17 00:00:00 2001
Message-Id: <d3ece29f674b85c055e15b955adc867e91e6b25e.1423872525.git.feitong.yi@intel.com>
In-Reply-To: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
References: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
From: "Thulasimani, Sivakumar" <sivakumar.thulasimani@intel.com>
Date: Wed, 28 Jan 2015 21:30:06 +0530
Subject: [PATCH 53/95] MUST_REBASE [VPG]: drivers/video/adf: hotplug and
 flicker WA

This patch fixes the hotplug notifications in intel_adf_interface.
Also temporarily worksaround bug in intel_adf_device that disables
primary plane when in dual display scenario.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed.

Issue: GMINL-5485
Change-Id: Id17230ae251e3e714c5153003f94f6a0fab98926
Signed-off-by: Thulasimani, Sivakumar <sivakumar.thulasimani@intel.com>
---
 drivers/video/adf/intel/intel_adf_device.c    |    6 +++++-
 drivers/video/adf/intel/intel_adf_interface.c |   19 +++++++++++++------
 2 files changed, 18 insertions(+), 7 deletions(-)

diff --git a/drivers/video/adf/intel/intel_adf_device.c b/drivers/video/adf/intel/intel_adf_device.c
index 5d8ceaa..b891ba8 100644
--- a/drivers/video/adf/intel/intel_adf_device.c
+++ b/drivers/video/adf/intel/intel_adf_device.c
@@ -458,7 +458,11 @@ static void disable_unused_overlay_engines(struct list_head *active_engs,
 		if (post_obj_set_find_obj(post_engs, eng))
 			continue;
 		/*disable this engine*/
-		eng->plane->ops->disable(eng->plane);
+		/*
+		 * FIXME: temp removed to avoid plane disable in dual display
+		 * scenarnario
+		 * eng->plane->ops->disable(eng->plane);
+		 */
 	}
 }
 
diff --git a/drivers/video/adf/intel/intel_adf_interface.c b/drivers/video/adf/intel/intel_adf_interface.c
index 71a5f17..1a5c51a 100644
--- a/drivers/video/adf/intel/intel_adf_interface.c
+++ b/drivers/video/adf/intel/intel_adf_interface.c
@@ -261,7 +261,8 @@ static int set_preferred_mode(struct intel_adf_interface *intf)
 		!pipe->ops->get_modelist)
 		return -EINVAL;
 
-	pipe->ops->get_modelist(pipe, &modelist, &n_modes);
+	pipe->ops->get_modelist(pipe, &intf->base.modelist,
+					&intf->base.n_modes);
 	pipe->ops->get_preferred_mode(pipe, &preferred_mode);
 
 
@@ -273,6 +274,9 @@ static int set_preferred_mode(struct intel_adf_interface *intf)
 	if (err)
 		goto out_err0;
 
+	modelist = intf->base.modelist;
+	n_modes = intf->base.n_modes;
+
 	/*
 	 * Only send HPD event in case of hot-pluggable displays but set
 	 * the interface detect status for every interface, as this will be
@@ -323,9 +327,13 @@ static void handle_hotplug_connected(struct intel_adf_interface *intf)
 	if (!pipe || !pipe->ops || !pipe->ops->is_screen_connected)
 		return;
 
-	if (pipe->ops->is_screen_connected(pipe))
+	if (pipe->ops->is_screen_connected(pipe)) {
+		/* fill the interface modelist */
+		pipe->ops->get_modelist(pipe, &intf->base.modelist,
+					&intf->base.n_modes);
 		queue_kthread_work(&intf->event_handling_worker,
 			&intf->hotplug_connected_work);
+	}
 }
 
 static void handle_hotplug_disconnected(struct intel_adf_interface *intf)
@@ -468,12 +476,11 @@ int intel_adf_interface_init(struct intel_adf_interface *intf,
 	if (err)
 		goto out_err3;
 
-	/* fill the interface modelist */
-	pipe->ops->get_modelist(pipe, &intf->base.modelist,
-				&intf->base.n_modes);
-
+		pipe->ops->get_modelist(pipe, &intf->base.modelist,
+					&intf->base.n_modes);
 	/*turn on this interface if screen was connected*/
 	if (pipe->ops->is_screen_connected(pipe)) {
+		/* fill the interface modelist */
 		err = set_preferred_mode(intf);
 		if (err) {
 			dev_err(dev->base.dev, "%s: failed to handle hotplug\n",
-- 
1.7.9.5

